### PR TITLE
Refactor selector peeking

### LINKED ISSUE #948: Refactor selector peeking in syscall handlers

### PR DESCRIPTION
**Pull request by drknzz**
Closes #948 

## Introduced changes

- 

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Review by Arcticae — COMMENTED** [review]
I think those methods are scattered right now, and could be centralized in one module. Apart from that i think it's a step in a good direction 👍

File: crates/cheatnet/src/execution/syscalls.rs
Comment (Arcticae) [inline crates/cheatnet/src/execution/syscalls.rs:1]: Perhaps we can pack it into a struct + a trait?

File: crates/forge/src/test_execution_syscall_handler.rs [hunk 1]
```diff
@@ -691,18 +679,11 @@ fn execute_syscall(
...
-    cheatable_syscall_handler
-        .child
-        .verify_syscall_ptr(system_ptr)?;
-
-    // We peek into memory to check the selector
-    let selector = DeprecatedSyscallSelector::try_from(felt_to_stark_felt(
-        &vm.get_integer(cheatable_syscall_handler.child.syscall_ptr)
-            .unwrap(),
-    ))?;
+    let selector = DeprecatedSyscallSelector::try_from(get_syscall_selector(
```
Comment (Arcticae) [inline crates/forge/src/test_execution_syscall_handler.rs:682]: sidenote: Why is this one `DeprecatedSyscallSelector` 🤔

**Comment by Arcticae** [comment]
@MaksymilianDemitraszek can you review before merging this?

File: crates/cheatnet/src/execution/syscalls.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    syscall_handler.verify_syscall_ptr(initial_syscall_ptr)?;
+    let selector = stark_felt_from_ptr_immutable(vm, &initial_syscall_ptr)?;
+    Ok(selector)
+}
+
+pub fn stark_felt_from_ptr_immutable(
+    vm: &VirtualMachine,
+    ptr: &Relocatable,
+) -> Result<StarkFelt, VirtualMachineError> {
+    Ok(felt_to_stark_felt(&felt_from_ptr_immutable(vm, ptr)?))
```
Comment (piotmag769) [inline crates/cheatnet/src/execution/syscalls.rs:30]: We can use our conversions trait here. Can we get rid of those chain call entirely here?

File: crates/cheatnet/src/execution/contract_execution_syscall_handler.rs [hunk 3]
```diff
@@ -56,14 +41,10 @@ impl HintProcessorLogic for ContractExecutionSyscallHandler<'_> {
...
-            let selector = std::str::from_utf8(selector).unwrap();
-            let inputs = match extract_input(vm, input_start, input_end) {
-                Ok(inputs) => inputs,
-                Err(err) => return Err(err),
-            };
+            let selector = parse_selector(selector)?;
+            let inputs = extract_input(vm, input_start, input_end)?;
 
-            match selector {
+            match selector.as_str() {
```
Comment (war-in) [inline crates/cheatnet/src/execution/contract_execution_syscall_handler.rs:47]: Wouldn't be better to return a string from `parse_selector`? I think in both cases we use `selector` as a string

**Comment by MaksymilianDemitraszek** [comment]
Would be nice if you included a sentence in the PR description, briefly describing the taken approach.

**Comment by MaksymilianDemitraszek** [comment]
Dead code: https://github.com/foundry-rs/starknet-foundry/blob/987b1aaa2d4da99624871405e0652c38871cf4c3/crates/cheatnet/src/execution/deprecated/syscalls.rs#L180

File: crates/forge/src/test_execution_syscall_handler.rs (hunk 1 above)
Comment (MaksymilianDemitraszek) [reply crates/forge/src/test_execution_syscall_handler.rs:682]: I think we should use SyscallSeclector here instead

File: crates/cheatnet/src/execution/syscalls.rs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+pub fn get_syscall_selector(
+    syscall: &ResOperand,
+    vm: &mut VirtualMachine,
+    syscall_handler: &SyscallHintProcessor<'_>,
+) -> Result<StarkFelt, HintError> {
+    let initial_syscall_ptr = get_ptr_from_res_operand_unchecked(vm, syscall);
+    syscall_handler.verify_syscall_ptr(initial_syscall_ptr)?;
+    let selector = stark_felt_from_ptr_immutable(vm, &initial_syscall_ptr)?;
+    Ok(selector)
+}
```
Comment (MaksymilianDemitraszek) [inline crates/cheatnet/src/execution/syscalls.rs:24]: Probably could do something like this (haven't tested it)
```
pub fn get_syscall_selector<T: TryFrom<StarkFelt>() -> T {
T::try_from(...)
}
```

File: crates/cheatnet/src/execution/cheatable_syscall_handler.rs [hunk 5]
```diff
@@ -165,12 +130,9 @@ impl CheatableSyscallHandler<'_> {
...
         hint: &StarknetHint,
     ) -> HintExecutionResult {
-        // We peek into the selector without incrementing the pointer as it is done later
         let syscall = get_syscall_operand(hint)?;
-        let initial_syscall_ptr = get_ptr_from_res_operand_unchecked(vm, syscall);
-        let selector =
-            SyscallSelector::try_from(stark_felt_from_ptr_immutable(vm, &initial_syscall_ptr)?)?;
-        self.verify_syscall_ptr(initial_syscall_ptr)?;
+        // We peek into the selector without incrementing the pointer as it is done later
+        let selector = SyscallSelector::try_from(get_syscall_selector(syscall, vm, &self.child)?)?;
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply crates/cheatnet/src/execution/cheatable_syscall_handler.rs:135]: (proposed generic solution in another comment)

**Comment by Arcticae** [comment]
Closing in favor of Extended Runtime refactor

**Pull request closed without merging**